streams/src/main/java/org/apache/kafka/streams/processor/internals/AssignedTasks.java

```diff
@@ -0,0 +1,319 @@
+        log.warn("{} Producer of {} {} fenced; closing zombie task", logPrefix, taskTypeName, task.id);
+        try {
+            task.close(false);
+        } catch (final Exception e) {
+            log.warn("{} Failed to close zombie {} due to {}, ignore and proceed", taskTypeName, logPrefix, e);
+        }
+    }
+
+    boolean hasRunningTasks() {
+        return !running.isEmpty();
+    }
+
+    boolean maybeResumeSuspendedTask(final TaskId taskId, final Set<TopicPartition> partitions) {
+        if (suspended.containsKey(taskId)) {
+            final T task = suspended.get(taskId);
+            if (task.partitions().equals(partitions)) {
+                suspended.remove(taskId);
+                log.trace("{} resuming suspended {} {}", logPrefix, taskTypeName, taskId);
+                task.resume();
+                transitionToRunning(task);
+                return true;
+            }
+        }
+        return false;
+    }
+
+    private void transitionToRunning(final T task) {
+        running.put(task.id(), task);
+        for (TopicPartition topicPartition : task.partitions()) {
+            runningByPartition.put(topicPartition, task);
+        }
+        for (TopicPartition topicPartition : task.changelogPartitions()) {
+            runningByPartition.put(topicPartition, task);
+        }
+    }
+
+    T runningTaskFor(final TopicPartition partition) {
+        return runningByPartition.get(partition);
+    }
+
+    Set<TaskId> runningTaskIds() {
+        return running.keySet();
+    }
+
+    Map<TaskId, T> runningTaskMap() {
+        return Collections.unmodifiableMap(running);
+    }
+
+    public String toString(final String indent) {
+        final StringBuilder builder = new StringBuilder();
+        describe(builder, running.values(), indent, "Running:");
+        describe(builder, suspended.values(), indent, "Suspended:");
+        describe(builder, restoring.values(), indent, "Restoring:");
+        describe(builder, created.values(), indent, "New:");
+        return builder.toString();
+    }
+
+    private void describe(final StringBuilder builder,
+                          final Collection<T> tasks,
+                          final String indent,
+                          final String name) {
+        builder.append(indent).append(name);
+        for (final T t : tasks) {
+            builder.append(indent).append(t.toString(indent + "\t\t"));
+        }
+        builder.append("\n");
+    }
+
+    List<AbstractTask> allInitializedTasks() {
+        final List<AbstractTask> tasks = new ArrayList<>();
+        tasks.addAll(running.values());
+        tasks.addAll(suspended.values());
+        tasks.addAll(restoring.values());
+        return tasks;
+    }
+
+    Collection<T> suspended() {
+        return suspended.values();
+    }
+
+    Collection<T> restoring() {
+        return Collections.unmodifiableCollection(restoring.values());
+    }
+
+    Collection<TaskId> allAssignedTaskIds() {
```

Comment:
i see no harm in having both. It makes it simpler for the calling code when you just want to have a list of taskIds rather than the tasks. Otherwise we will have the same blocks of code extracting taskIds in various locations.